### ISSUE TITLE
PQS.sphere.sx structure is non-linear

### ISSUE THREAD

**Issue by R-T-B** [KSPModdingLibs/KSPCommunityFixes#41]
The bug can be summarized as follows:

PQS.sphere.sx is supposed to be a variable of linear structure spanning from 0 to 1 indicating a longitude point (similar to how PQS.sphere.sy functions properly for latitude).  It does not.  Range 0-0.75 behaves normally, but the final quadrant of longitude is in fact migrated nonlinearly into the -0.25 to 0 part of the value.  This causes obvious issues, and the root cause is not only diffilcult to diagnose, but if truly corrected at the source probably would break other game facilities that have long worked around this bug.

The only known bug from this is that Kopernicus LongitudeRange values ignore the final quadrant, as a direct result of this nonlinear value representation.  A patch can be made to fix this in PQSLandControl.  In that class, there is this line:

`this.vLon = this.sphere.sx + (double)this.longitudeBlend * this.longitudeSimplex.noise(data.directionFromCenter);`

I propose that line be replaced with:

`double correctedSphereSx;
        if (this.sphere.sx > 0.0)
        {
            correctedSphereSx = this.sphere.sx;
        }
        else
        {
            correctedSphereSx = this.sphere.sx + 1.0;
        }
        this.vLon = correctedSphereSx + (double)this.longitudeBlend * this.longitudeSimplex.noise(data.directionFromCenter);`

That should work around the nonlinearity in the one known symptomatic instance, while not mucking about with any stock systems that may already work around the core bug.  Feel free to provide your thoughts.

**Comment by gotmachine** [comment]
So, to be clear, we are talking about `PQSLandControl.OnVertexBuildHeight()` ?

This piece of code is indeed quite hard to follow. This pattern of having instance fields being used as local variables shared by multiple methods all over the place is really awful.

Despite this being quite weird, I think `sx` is designed to be in the `[-0.25, 0.75]` range. There are way too many places where this is used for this to be a mistake. Since this is mapping to texture UV coordinates, my interpretation is that there is a deliberate intent to have an offset between the texture origin and the 0° longitude. 

And in fact, when using those values as UV coordinates, the stock code always does `val = Math.Abs(val - Math.Floor(val))` which is equivalent as your modified code and transform back the value to a valid `[0, 1]` UV range.

I think the actual mistake is indeed `PQSLandControl.OnVertexBuildHeight()` trying to get back a longitude value from the UV coordinates without accounting for that offset, so your fix makes sense from that PoV.

I will implement that fix and do a release ASAP.

**Comment by R-T-B** [comment]
Yes you got the right function.  And looking at it I agree with everything you say...  just seems very odd a choice of coordinate system.

**Comment by R-T-B** [comment]
Forgot to answer one point:  LandControl can do a bit more than scatters.  It can do some terrain coloring effects amongst other things.
